keps/sig-storage/20190204-kubernetes-csi-release-process.md

```diff
@@ -0,0 +1,233 @@
+    - [Drawbacks](#drawbacks)
+    - [Alternatives](#alternatives)
+    - [Infrastructure Needed](#infrastructure-needed)
+
+
+## Summary
+
+The Kubernetes [Storage
+SIG](https://github.com/kubernetes/community/tree/master/sig-storage)
+maintains a set of components under the
+[kubernetes-csi](https://github.com/kubernetes-csi) GitHub
+organization. Those components are intentionally not part of core
+Kubernetes even though they are maintained by the Kubernetes project.
+
+This document explains how these components are released.
+
+## Motivation
+
+So far, the process for tagging and building components has been
+fairly manual, with some assistance by Travis CI. There has been no
+automatic end-to-end testing.
+
+With CSI support reaching GA in Kubernetes 1.13, it is time to define
+and follow a less work-intensive and more predictable approach.
+
+### Goals
+
+- define what a "kubernetes-csi release" is
+- unit testing as mandatory pre-submit check for all components 
+- E2E testing as mandatory pre-submit check for components that get
+  deployed in a cluster
+- define a release process for each component and a "kubernetes-csi release"
+
+### Non-Goals
+
+- change responsibilities among the kubernetes-csi maintainers - as before, each
+  component will have some main maintainer who is responsible for releasing updates
+  of that component
+- automatically generate release notes
+- automatically generate "kubernetes-csi release" artifacts - this will need further thoughts
+
+
+## Proposal
+
+### Versionioning
+
+Each of the components has its own documentation, versioning and
+release notes. [Semantic versioning](https://semver.org/) is used for
+components that provide a stable API, like for example:
+- external-attacher
+- external-provisioner
+- external-snapshotter
+- cluster-driver-registrar
+- node-driver-registrar
+- csi-driver-host-path
+- csi-lib-utils
+
+Other components are internal utility packages that are
+not getting tagged:
+- csi-release-tools
+
+A "kubernetes-csi release" is a specific set of component
+releases. It's also called the "combined release". Documentation for
+combined releases is found on [kubernetes-csi
+docs](https://kubernetes-csi.github.io/docs/). The combined release is
+typically going to be prepared and announced less frequently than the
+individual component releases. It is therefore only a recommendation
+that downstream users use this combination of components. In
+particular, individual components might also get minor updates after a
+kubernetes-csi release without updating that combined release.
+
+The
+[hostpath example deployment](https://github.com/kubernetes-csi/csi-driver-host-path/tree/master/deploy)
+defines the components that are part of a combined release. This
+implies that the hostpath driver repo must be updated and tagged to
+create a new combined release. Therefore the hostpath driver's version
+becomes the kubernetes-csi release version, which is increased
```

Comment:
What happens if we need to bump the hostpath driver for driver-specific changes/bug fixes, and not necessarily for updating the sidecars?

Reply:
Then we do another "kubernetes-csi release" with a version bump on the hostpath driver repo. That new release then contains that hostpath driver change.

It's not possible to release an update of the hostpath driver without also doing a new kubernetes-csi release - it's special in that regard.

If we don't want that, then we need to have a separate repo which defines the current set of sidecars that are in the kubernetes-csi release. That repo then should also have the example hostpath deployment, because that is effectively that definition.

Reply:
Then we do another "kubernetes-csi release" with a version bump on the hostpath driver repo. That new release then contains that hostpath driver change.

It's not possible to release an update of the hostpath driver without also doing a new kubernetes-csi release - it's special in that regard.

If we don't want that, then we need to have a separate repo which defines the current set of sidecars that are in the kubernetes-csi release. That repo then should also have the example hostpath deployment, because that is effectively that definition.

Reply:
I think it's a odd that a driver is being used as the canonical place to define a combined release.

Can we just point to the hostpath example at a specific tag in the docs?  Do we actually need an official release number?  It could be confusing because the sidecars will all have their own version numbers that may not match.

Reply:
I myself have never been a fan of moving the example deployment into the csi-driver-host-path repo because of the implications (must do new release of driver whenever a sidecar changes), but that's where we agreed to put the example deployment.

At least that example actually does get tested. One other example deployment has been completely broken for a while until eventually someone noticed (https://github.com/kubernetes-csi/external-attacher/issues/120).

> Can we just point to the hostpath example at a specific tag in the docs?

The example is a set of YAML files and a script: https://github.com/kubernetes-csi/csi-driver-host-path/tree/master/deploy

Are you suggesting that the YAML files shouldn't contain version numbers and instead get that information from elsewhere? That may work, but what if the rest of the YAML files needs to be changed? Then we are still stuck with updating the csi-driver-host-path.

> Do we actually need an official release number?

Do we still have the concept of doing a coordinated update of all components? If yes, what do we call that update ("kubernetes-csi release"?) and how do we distinguish between different updates (the one from "February 2019" has these changes since the release from "January 18th 2019")?

Let's suppose we revisit the decision to have the example deployment in the hostpath repo. The alternative is this:
- move the example to a new kubernetes-csi/csi-release repo
- cut a new release of the hostpath driver if (and only if) the driver itself has changed, i.e. treat it like any of the other components
- tagging and testing of the csi-release repo defines what's in a "kubernetes-csi release", with its own version numbering

Reply:
I'm not actually sure we need a concept of a coordinated release.  Right now we unofficially align releases with the Kubernetes release, but I think that's mainly due to the release process being mostly manual at the moment.  Once we have an automated release process, I think we can release more frequently and independently.

What I'm trying to suggest is that we don't have an "official" bundle of sidecars and versions, nor an official complete example.  We can point to the hostpath driver as a complete example that people can reference as a starting point, but it's just a suggestion because hostpath may not be able to test every single feature we support in all the sidecars (although we should strive to maintain and test it as much as we can).

Reply:
That means that kubernetes-csi drops its own "last line of defense" against a botched component release, because once such a release is tagged and built, it is by definition the latest release of that component and downstream users are invited to pick it up. I'm fine with that, because after the discussion in https://github.com/kubernetes-csi/csi-release-tools/issues/6 I expect that these newly released images do not end up on production clusters immediately, but rather get vetted as part of  whatever release and deployment model the actual CSI drivers use.